File: behaviorial/template-method/typescript/concrete.ts
import { ReportGenerator } from './abstract';

export class PDFReportGenerator extends ReportGenerator {
	protected collectData(): void {
		console.log('Collecting data for PDF report...');
	}

	protected processData(): void {
		console.log('Processing data for PDF report...');
	}

	protected formatData(): void {
		console.log('Formatting data for PDF report as PDF...');
	}
}

export class ExcelReportGenerator extends ReportGenerator {
	protected collectData(): void {
		console.log('Collecting data for Excel report...');
	}

	protected processData(): void {
		console.log('Processing data for Excel report...');
	}

	protected formatData(): void {
		console.log('Formatting data for Excel report as Excel...');
	}
}

File: behaviorial/template-method/python/concrete.py
from abstract import ReportGenerator


class PDFReportGenerator(ReportGenerator):
    def collect_data(self):
        print("Collecting data for PDF report...")

    def process_data(self):
        print("Processing data for PDF report...")

    def format_data(self):
        print("Formatting data for PDF report as PDF...")


class ExcelReportGenerator(ReportGenerator):
    def collect_data(self):
        print("Collecting data for Excel report...")

    def process_data(self):
        print("Processing data for Excel report...")

    def format_data(self):
        print("Formatting data for Excel report as Excel...")
